src/spanningtrees/boruvka.jl

```diff
@@ -0,0 +1,56 @@
+@traitfn function boruvka_mst(graph::AG::(!IsDirected), 
+    distmx::AbstractMatrix{T} = weights(graph); 
+    minimize = true) where {T<:Real,U,AG<:AbstractGraph{U}}
+    
+    f_distmx = minimize ? copy(distmx) : copy(-distmx)
+
+    djSet = IntDisjointSets(nv(graph))
+
+    mst = Vector{edgetype(graph)}()
+    sizehint!(mst, nv(graph) - 1)
+
+    weight = 0.0
+    while (num_groups(djSet) > 1)
+
+        cheapest = Array{Union{edgetype(graph), Bool}}(undef, nv(graph))
+
+        for edge in edges(graph)
+            set1 = find_root(djSet, src(edge))
+            set2 = find_root(djSet, dst(edge))
+            if (set1 != set2)
+                e1 = cheapest[set1]
+                if (cheapest[set1] == false || f_distmx[src(e1), dst(e1)] > f_distmx[src(edge), dst(edge)])
```

Comment:
Is it possible that `cheapest[set1]` refers to undefined memory at this point? How do you know it will be `false`?
also: since you're assigning `cheapest[set1]` to `e1` in the previous line, it might be clearer to refer to this as `e1` here.

Reply:
> Is it possible that `cheapest[set1]` refers to undefined memory at this point? How do you know it will be `false`?
> also: since you're assigning `cheapest[set1]` to `e1` in the previous line, it might be clearer to refer to this as `e1` here.

No, it's not possible cause previously I've defined the length of the array ( I will change to Vector) is nv(graph), and it is impossible to have a number of sets greater than the number of vertices

Reply:
in the first run through the loop on line 17, you have not set any values in `cheapest`. Therefore, `e1` will refer to undefined memory in line 21 (from the allocation on lline 15).

You're probably getting lucky so far that it's using a zeroed-out page for allocation on L17, but there's no guarantee of this on other systems or even on your system.

Am I missing something?

Reply:
> in the first run through the loop on line 17, you have not set any values in `cheapest`. Therefore, `e1` will refer to undefined memory in line 21 (from the allocation on lline 15).
> 
> You're probably getting lucky so far that it's using a zeroed-out page for allocation on L17, but there's no guarantee of this on other systems or even on your system.
> 
> Am I missing something? Here's an example of what happens when you allocate "raw" memory:
> 
> ```
> julia> a = Vector{Int}(undef, 4)
> 4-element Array{Int64,1}:
>  4596247184
>  4453068800
>  4452350848
>           0
> ```
> 
> when you then reference `a[2]` before assigning to it, you'll get `4453068800`. This is almost certainly not what you want to do.

when you initiallize it as a Union with false or nothing it will always initiallize the whole with false or nothing

Reply:
> when you initiallize it as a Union with false or nothing it will always initiallize the whole with false or nothing

Where is that documented? I’ve not heard that before.

Reply:
> > when you initiallize it as a Union with false or nothing it will always initiallize the whole with false or nothing
> 
> Where is that documented? I’ve not heard that before.

I`m going to change this line to : 
cheapest = Vector{Union{edgetype(graph), Nothing}}(nothing, nv(graph))
I could see that in this part of documentation:
https://docs.julialang.org/en/v1/base/arrays/
to make it easier to find, try using Ctrl+F on " Array{T,N}(nothing, dims) "  I saw it in the examples

Reply:
> I`m going to change this line to : `cheapest = Vector{Union{edgetype(graph), Nothing}}(nothing, nv(graph))`

That's safer. In general, unless you're going to be filling the vector immediately or know that you're only going to access elements after you assign to them, you should avoid `undef`.